### PR TITLE
Compatibility policy

### PR DESCRIPTION
**Pull request by yrodiere**
**This is just my take on what the policy currently is, there may be incorrect information in here**

I took a wiki page by Emmanuel from 2014 ([here](https://github.com/hibernate/hibernate-orm/wiki/Compatibility-Considerations)), updated it a bit and ported it to asciidoc, then added a few sections about Hibernate Search + Lucene and Hibernate Search + Elasticsearch (which is what I was really after).

I also added links to the compatibility policy from the compatibility matrixes we have in all projects.

[Here is the result](https://staging.hibernate.org/community/compatibility-policy/).

@gsmet @Sanne would you mind proof-reading it?

### CODE REVIEW

File: community/compatibility-policy.adoc [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+For example: `4.0.0.Final`, `5.2.0.Beta2`, `5.4.0.CR1`.
+
+For the most part we follow the guidelines as defined by
+https://community.jboss.org/wiki/JBossProjectVersioning[JBoss Project Versioning Guidelines],
+so that is a good background material.
+
+Hibernate, however, applies some additional semantics. Roughly speaking:
+
+* Major updates imply changes to existing APIs.
+* Minor updates simply simply additions to existing APIs.
```
Comment (gsmet) [inline community/compatibility-policy.adoc]:
```suggestion
* Minor updates simply additions to existing APIs.
```

**Review by gsmet — DISMISSED** [review]
I wonder if we should have a link to the various compatibility matrix and an explanation that versions for projects are entirely independant.

**Comment by yrodiere** [comment]
Right, will do.

File: community/compatibility-policy.adoc (hunk 1 above)
Comment (sebersole) [reply community/compatibility-policy.adoc]: `* Minor updates simply additions to existing APIs.` is not not proper grammar.  And I'm assuming "imply" was meant again:

`* Minor updates imply additions to existing APIs.`

Also, might be good idea to highlight that changes as well as removals fall under "changes".  I think we all understand that, but just to be clear.

**Comment by sebersole** [comment]
Personally I'd like to highlight `@Internal` as a means for categorizing something as internal.  Also, similarly, `@Incubating`

**Comment by yrodiere** [comment]
> Personally I'd like to highlight @Internal as a means for categorizing something as internal. Also, similarly, @Incubating

Done

**Comment by Sanne** [comment]
I totally love that you took the time to port this information to where it belongs.

My only doubt is in relation to the Jakarta EE plan we have: we might not end up doing it in 5.5, but the possibility is non-negligible. Should we explicitly call it out as an example of things we might need doing? There is the red box stating "While we strive to follow them as much as possible, there are times when exceptions need to be made.", but clearly we dont strictly have the *need* to make such a change in 5.5, we could have called it 6.0 if that rule was to be taken more seriously.

Alternatively, we could relax the language a little by stating that we consider various factors, including the difficulty of people to actually migrate:
I believe we do strive to respect the expectation that it's easier to migrate a minor than it is to upgrade a major.

**Comment by yrodiere** [comment]
> Alternatively, we could relax the language a little by stating that we consider various factors, including the difficulty of people to > actually migrate:
> I believe we do strive to respect the expectation that it's easier to migrate a minor than it is to upgrade a major.

I'll add an example stating this:

> Another exception is when backporting support for standards. We might decide to introduce "transition" minors that upgrade to a newer version of a standard while keeping backward compatibility everywhere else, in order to ease the migration to the next major version. For example Hibernate ORM 5.5 might upgrade to Jakarata 9's persistence API (and change nothing else) to ease the migration to Hibernate ORM 6.0.

@Sanne Would that work? Feel free to reword if it doesn't.

**Comment by Sanne** [comment]
@yrodiere perfect. Just a typo in "Jakarata".. Thanks!

**Comment by yrodiere** [comment]
Merged. Thanks for the reviews!

**Pull request merged**